Docs: feature documentation for `build.jobs` (#9138)
readthedocs/readthedocs.org: docs/user/build-customization.rst

Build customization
===================

> Read the Docs has a :ref:`well-defined build process <builds>` that may not be enough for all the use cases.

Reviewer: ```suggestion
Read the Docs has a :ref:`well-defined build process <builds>` that works for many projects, 
but we offer additional customization to support more uses of our platform.
```

> This page explains how to customize Read the Docs build process to get the most of it and support custom build processes.

Reviewer: ```suggestion
This page explains how to customize the build process to get the most of it and support custom build steps.
```

Another work for `jobs` or `steps` here. We really need to tighten this up, so users actually know what we're talking about. 

Is it:

* Build process: Total series of steps we take to build a set of content (wording gets way trickier here when we're doing non-docs projects...)
* Build steps: Each step of the process, specifically the ones that we run? 
* Build jobs: Non-standard steps that users run? 

@agjohnson thoughts here?

Reviewer 2: So, that sort of matches how I would describe it, but I also feel like it's a little inaccurate :upside_down_face: 

There are discrete `pre_install` and `post_install` jobs, and I see a future with an `install` job too. So I suppose we'd call those "jobs", and "steps" are the collection of those "jobs".

So, in :snake: :

```python
BuildProcess(steps=[
    InstallStep(jobs=[
        PreInstallJob(commands=[]),
        InstallJob(commands=[]),
        PostInstallJob(commands=[]),
    ]),
    BuildStep(jobs=[
        PreBuildJob(commands=[]),
        BuildJob(commands=[]),
        PostBuildJob(commands=[]),
    ]),
    ...
])
```

Perhaps pre/post install jobs are currently "custom jobs" or "user-defined jobs"? Or `build.jobs.install` is a "pre-defined job".

Author: > "custom jobs" or "user-defined jobs"? Or build.jobs.install is a "pre-defined job"

This is what I had in mind as well. Maybe I got confusing while writing the document 😅 , but I think it's what I wanted to describe: pre-defined jobs and user-defined jobs.

Author: I'm accepting Anthony's suggestion for now, but we can update it later if we find better wording.

Reviewer: I really like Anthony's code snippet. I'd argue that could be useful to put in the docs, but maybe too codey?

> In the normal build process the jobs ``checkout``, ``system_dependencies``, ``create_environment``, ``install``, ``build`` and ``upload`` are executed.

Author: I named them jobs here on purpose to be able to introduce `build.jobs` later in the doc. I want some way to link and relate "pre-defined jobs" with "user-defined jobs", so I think it'd be good to find a clear way to use jobs here if possible.

What about using "pre-defined jobs" as you mentioned before?

```suggestion
In the normal build process the pre-defined jobs ``checkout``, ``system_dependencies``, ``create_environment``, ``install``, ``build`` and ``upload`` are executed.
```

> In the normal build process the jobs ``checkout``, ``system_dependencies``, ``create_environment``, ``install``, ``build`` and ``upload`` are executed.

Reviewer 2: That still makes sense with the nomenclature of process/step/jobs :+1:

However, Read the Docs exposes in-between jobs to users so they can customize the build process by running shell commands.

> These in-between jobs are:

Author: OK, we will need to change the previous sentence as well.

:post_checkout:

:pre_system_dependencies:

:post_system_dependencies:

:pre_create_environment:

:post_create_environment:

:pre_install:

:post_install:

:pre_build:

> :post_build:

Reviewer: This renders pretty oddly. I think we just want a bulleted list?

Author: Yeah, I'm happy to improve the rendering of this. I was about to add content on each job but then realized that it would be too repetitive and useless. I'll try Anthony's suggestion with the table.

> These jobs can be declared by using a :ref:`config-file/index` with the :ref:`config-file/v2:build.jobs` key on it.

Reviewer: ```suggestion
These jobs can be declared by using a :doc:`/config-file/index` with the :ref:`config-file/v2:build.jobs` key on it.
```

The index link breaks here.

Let's say the project requires a mandatory command to be executed *before* installing any dependency into the Python environment and *after* the build has finished.
In that case, a config file similar to this one can be used:

.. code:: yaml

   version: 2
   build:
     os: "ubuntu-20.04"
     tools:
       python: "3.10"
     jobs:
       pre_install:
         - echo "Executing pre_install step..."
         - bash ./scripts/pre_install.sh
       post_build:
         - echo "Executing post_build step..."
         - curl -X POST \
           -F "project=${READTHEDOCS_PROJECT}" \

>            -F "version=${READTHEDOCS_VERSION}" https://my.company.com/webhooks/readthedocs/

Author: ```suggestion
           -F "version=${READTHEDOCS_VERSION}" https://example.com/webhooks/readthedocs/
```

.. note::

>    Currently, the default jobs (e.g. ``checkout``, ``system_dependencies``, etc) executed by Read the Docs are not possible to overwritte or skip.

Reviewer: ```suggestion
   Currently, the default jobs (e.g. ``checkout``, ``system_dependencies``, etc) executed by Read the Docs are not possible to override or skip.
```

> 

Author: Good point! I added some small bullets at this point for now to communicate this. We can expand this on its own section if we consider it necesary.

> 

Reviewer 2: That's good for now, we can come back and expand on that. There might be a good deal of explanation we can add here. I reviewed and put up some copy edits.

Examples
--------

> This is a non-exahustive list of common examples where using :ref:`config-file/v2:build.jobs` will be useful.
> Also note that these examples may require some adaption for each projects' use case.
> They are listed here just as a guide.

Reviewer: ```suggestion
We've included some common examples where using :ref:`config-file/v2:build.jobs` will be useful.
These examples may require some adaption for each projects' use case,
we recommend you use them as a starting point.
```

Unshallow clone
~~~~~~~~~~~~~~~

> Read the Docs does not perform a full clone on ``checkout`` step to reduce network data and speed up the build process.

Reviewer: Do we document this anywhere? Would be useful to link if we do.

Author: I think we don't have documented that we do a shallow clone, no.

Because of this reason, if there are extensions in the build process that depend on the full history, they will fail.
To avoid this, it's possible to unshallow the clone done by Read the Docs:

.. code:: yaml

   version: 2
   build:
     os: "ubuntu-20.04"
     tools:
       python: "3.10"
     jobs:
       post_checkout:
         - git fetch --unshallow

Generate documentation from annotated sources with Doxygen
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

It's possible to run Doxygen as part of the build process to generate documentation from annotated sources:

.. code:: yaml

   version: 2
   build:
     os: "ubuntu-20.04"
     tools:
       python: "3.10"
     jobs:

>        pre_build:
>          - doxygen

Reviewer: ```suggestion
       pre_build:
       # Note that this HTML won't be automatically uploaded, 
       # unless your documentation build includes it somehow. 
         - doxygen
```

Author: What this means? That they have to use a Sphinx config to include the HTML generated by doxygen?

Reviewer: Yes -- otherwise this expand doesn't really do anything.

Use MkDocs extensions with extra required steps
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

There are some MkDocs extensions that require specific commands to be run to generate extra pages before performing the build.
For example, `pydoc-markdown <http://niklasrosenstein.github.io/pydoc-markdown/>`_

.. code:: yaml

   version: 2
   build:
     os: "ubuntu-20.04"
     tools:
       python: "3.10"
     jobs:
       pre_build:
         - pydoc-markdown --build --site-dir "$PWD/_build/html"

Avoid having a dirty ``git`` index
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

> Read the Docs needs to modify some files before performing the build to be able to integrate with some of its features.
> Because of this reason, it could happen the git index gets dirty (e.i. it will detect modified files).
> In case this happens and the project is using any kind of extension that generates a version based on git metadata (like `setuptools_scm <https://github.com/pypa/setuptools_scm/>`_),
> this could case a miss generated version number.
> In that case, the git index can be updated to ignore the files that Read the Docs has modified.

Reviewer: ```suggestion
Read the Docs needs to modify some files before performing the build to be able to integrate with some of its features.
Because of this reason, it could happen the git index gets dirty (it will detect modified files).
In case this happens and the project is using any kind of extension that generates a version based on git metadata (like `setuptools_scm <https://github.com/pypa/setuptools_scm/>`_),
this could cause an invalid version number to be generated.
In that case, the git index can be updated to ignore the files that Read the Docs has modified.
```

.. code:: yaml

   version: 2
   build:
     os: "ubuntu-20.04"
     tools:
       python: "3.10"
     jobs:
       pre_install:

>          - git update-index --assume-unchanged environment.yml docs/conf.py

Reviewer: Interesting!

Build customization
===================

Read the Docs has a :ref:`well-defined build process <builds>` that works for many projects,
but we offer additional customization to support more uses of our platform.
This page explains how to extend the build process, using user-defined jobs to execute custom commands.

In the normal build process,
the pre-defined jobs ``checkout``, ``system_dependencies``, ``create_environment``, ``install``, ``build`` and ``upload`` are executed.
However, Read the Docs exposes extra jobs to users so they can customize the build process by running shell commands.
These extra jobs are:

.. list-table::
   :header-rows: 1

   * - Step
     - Customizable jobs
   * - Checkout
     - ``post_checkout``
   * - System dependencies
     - ``pre_system_dependencies``, ``post_system_dependencies``
   * - Create environment
     - ``pre_create_environment``, ``post_create_environment``
   * - Install
     - ``pre_install``, ``post_install``
   * - Build
     - ``pre_build``, ``post_build``
   * - Upload
     - There is no customizable jobs currently

.. note::

>    Currently, the default jobs (e.g. ``checkout``, ``system_dependencies``, etc) executed by Read the Docs are not possible to override or skip.

Reviewer 2: ```suggestion
   Currently, the pre-defined jobs (``checkout``, ``system_dependencies``, etc) executed by Read the Docs cannot be overridden or skipped.
```

These jobs can be declared by using a :doc:`/config-file/index` with the :ref:`config-file/v2:build.jobs` key on it.
Let's say the project requires commands to be executed *before* installing any dependency into the Python environment and *after* the build has finished.
In that case, a config file similar to this one can be used:

.. code:: yaml

   version: 2
   build:
     os: "ubuntu-20.04"
     tools:
       python: "3.10"
     jobs:
       pre_install:
         - bash ./scripts/pre_install.sh
       post_build:
         - curl -X POST \
           -F "project=${READTHEDOCS_PROJECT}" \
           -F "version=${READTHEDOCS_VERSION}" https://example.com/webhooks/readthedocs/

There are some caveats to knowing when using user-defined jobs:

> * ``PWD`` environment variable is the root of the cloned repository

Reviewer 2: ```suggestion
* The current working directory is at the root of your project's cloned repository
```

* Environment variables are expanded in the commands (see :doc:`environment-variables`)
* Each command is executed in a new shell process, so modifications done to the shell environment are not persistent between commands

> * Any command returning non-zero exit code makes the build to fail immediately

Reviewer 2: ```suggestion
* Any command returning non-zero exit code will cause the build to fail immediately
```

Because of this, extensions that depend on the full Git history will fail.
To avoid this, it's possible to unshallow the clone done by Read the Docs:

.. code:: yaml

   version: 2
   build:
     os: "ubuntu-20.04"
     tools:
       python: "3.10"
     jobs:
       post_checkout:
         - git fetch --unshallow

Generate documentation from annotated sources with Doxygen
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

It's possible to run Doxygen as part of the build process to generate documentation from annotated sources:

.. code:: yaml

   version: 2
   build:
     os: "ubuntu-20.04"
     tools:
       python: "3.10"
     jobs:
       pre_build:
         - doxygen

Use MkDocs extensions with extra required steps
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

There are some MkDocs extensions that require specific commands to be run to generate extra pages before performing the build.
For example, `pydoc-markdown <http://niklasrosenstein.github.io/pydoc-markdown/>`_

.. code:: yaml

   version: 2
   build:
     os: "ubuntu-20.04"
     tools:
       python: "3.10"
     jobs:
       pre_build:
         - pydoc-markdown --build --site-dir "$PWD/_build/html"

> Avoid having a dirty ``git`` index
> ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

Reviewer 2: To match other instances of "Git", but also, keep literal strings out of headings/toc

```suggestion
Avoid having a dirty Git index
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```
